clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -2867,6 +2913,163 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::next_record(output_ordinal_t outp
+
+        trace_marker_type_t marker_type;
+        uintptr_t marker_value;
+        bool is_marker =
+            record_type_is_marker(to_return.record, marker_type, marker_value);
+
+        bool indirect_branch_instr_pending = false;
+        bool branch_target_marker_pending = false;
+
+        if (is_marker && marker_type == TRACE_MARKER_TYPE_BRANCH_TARGET &&
+            marker_value == output_info_t::BRANCH_TARGET_NEEDS_FIXING) {
+            branch_target_marker_pending = true;
+        } else if (is_indirect_branch_instr && has_indirect_branch_target &&
+                   indirect_branch_target == output_info_t::BRANCH_TARGET_NEEDS_FIXING) {
+            indirect_branch_instr_pending = true;
+        }
+        if (branch_target_marker_pending || indirect_branch_instr_pending) {
+            // We need to read-ahead trace records until we know the next pc of the
+            // branch.
+            addr_t next_pc = 0;
+            VPRINT(this, 2, "next_record[%d]: Reading ahead to find target pc for %s\n",
+                   output,
+                   (branch_target_marker_pending ? "branch_target marker"
+                                                 : "indirect branch instr"));
+            while (next_pc == 0) {
+                stream_status_t res =
+                    next_record_internal(output, next_record, next_input, cur_time);
+                if (res == sched_type_t::STATUS_EOF) {
+                    // We let next_pc remain zero, to indicate that the indirect branch
+                    // instruction should be removed.
+                    break;
+                } else if (res != sched_type_t::STATUS_OK) {
+                    // If there's a non-OK response before we figure out the next pc, pass
+                    // that on. We'll retry later if this was simply because the output
+                    // stream isn't ready to produce the next record yet. The queued up
+                    // records will also be returned at that point.
+                    return res;
+                } else {
+                    finalized_record_t read_ahead_record(
+                        next_record, next_input,
+                        outputs_[output].in_context_switch_code ||
+                            outputs_[output].in_syscall_code);
+                    outputs_[output].kibt_readahead_q.push(read_ahead_record);
+
+                    addr_t maybe_next_pc = 0;
+                    // We want the instr after the one corresponding to the indirect
+                    // branch target we're looking for. This could be the next user-space
+                    // instruction after the kernel sequence ends, or for
+                    // syscall-to-switch it may be the first instruction of the switch
+                    // sequence.
+                    if (outputs_[output].kibt_readahead_q.instrs_count() == 2 &&
+                        record_type_is_instr(next_record, &maybe_next_pc)) {
```

Comment:
Do we need the ==2 check? Shouldn't we stop on *any* instr? I think the ==2 should be moved to an assert inside an is-instr if().

Reply:
> Shouldn't we stop on any instr?

No, for the record_scheduler_t, the next instr we'll see after the branch_target marker would be the indirect branch instr corresponding to the marker. We want to skip that.

For both memref_t and trace_entry_t cases, the target instr of the indirect branch instr would be at instr_count==2 in the queue.